domain/user/state/state.go

```diff
@@ -0,0 +1,335 @@
+// users creator is set and does not exist then an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUser(ctx context.Context, uuid domainuser.UUID, user user.User, creatorUUID domainuser.UUID) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		return addUser(ctx, tx, uuid, user, creatorUUID)
+	})
+
+	return err
+}
+
+// AddUserWithPasswordHash will add a new user to the database with the
+// provided password hash and salt. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. If the users creator
+// does not exist or has been previously removed an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUserWithPasswordHash(
+	ctx context.Context,
+	uuid domainuser.UUID,
+	user user.User,
+	creatorUUID domainuser.UUID,
+	passwordHash string,
+	salt []byte,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = addUser(ctx, tx, uuid, user, creatorUUID)
+		if err != nil {
+			return err
+		}
+
+		return setPasswordHash(ctx, tx, uuid, passwordHash, salt)
+	})
+
+	return err
+}
+
+// AddUserWithActivationKey will add a new user to the database with the
+// provided activation key. If the user already exists an error that
+// satisfies usererrors.AlreadyExists will be returned. if the users creator
+// does not exist or has been previously removed an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUserWithActivationKey(
+	ctx context.Context,
+	uuid domainuser.UUID,
+	user user.User,
+	creatorUUID domainuser.UUID,
+	activationKey string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err = addUser(ctx, tx, uuid, user, creatorUUID)
+		if err != nil {
+			return err
+		}
+		return setActivationKey(ctx, tx, uuid, activationKey)
+	})
+	if err != nil {
+		return fmt.Errorf("cannot add user %q with activation key: %w", user.Name, err)
+	}
+
+	return nil
+}
+
+// GetUser will retrieve the user specified by name from the database where
+// the user is active and has not been removed. If the user does not exist
+// or is deleted an error that satisfies usererrors.NotFound will be
+// returned.
+func (st *State) GetUser(ctx context.Context, userUUID domainuser.UUID) (domainuser.User, error) {
+	db, err := st.DB()
+	if err != nil {
+		return domainuser.User{}, fmt.Errorf("cannot get DB access: %w", err)
+	}
+
+	var domainUser domainuser.User
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		domainUser, err = getUser(ctx, tx, userUUID)
```

Comment:
is `getUser` being used by anything else other then this function? Should we just integrate it back in?

Reply:
Yep, for now, it is used only here, but I wish to use the same template as we did for `addUser` in case we wish to use `getUser` in complex transactions in the future.